File: src/platform_app/modules/identity/ports.py
"""Replaceable outbound identity interfaces."""

from datetime import datetime
from typing import Protocol


class EmailCodeSender(Protocol):
    def send_code(self, email: str, code: str, expires_at: datetime) -> None:
        """Deliver a single-use university-email verification code."""


class InMemoryEmailCodeSender:
    """Demo adapter that retains codes in process without logging them."""

    def __init__(self) -> None:
        self._outbox: dict[str, tuple[str, datetime]] = {}

    def send_code(self, email: str, code: str, expires_at: datetime) -> None:
        self._outbox[email] = (code, expires_at)

    def code_for(self, email: str) -> str:
        return self._outbox[email][0]

File: api/_src/modules/identity/ports.ts
/**
 * Replaceable outbound identity interfaces (port of identity.ports).
 */

export interface EmailCodeSender {
  /** Deliver a single-use university-email verification code. */
  sendCode(email: string, code: string, expiresAt: Date): void | Promise<void>;
}

/**
 * Demo adapter that retains codes in process without logging them. Also backs
 * the optional development inbox so the browser demo can display the code
 * (there is no real email integration in the demo).
 */
export class InMemoryEmailCodeSender implements EmailCodeSender {
  private readonly outbox = new Map<string, { code: string; expiresAt: Date }>();

  sendCode(email: string, code: string, expiresAt: Date): void {
    this.outbox.set(email, { code, expiresAt });
  }

  /** The most recent code for an address. Throws if none was issued. */
  codeFor(email: string): string {
    const entry = this.outbox.get(email);
    if (entry === undefined) {
      throw new Error(`no verification code issued for ${email}`);
    }
    return entry.code;
  }

  /** The most recent code for an address, or null. */
  peek(email: string): { code: string; expiresAt: Date } | null {
    return this.outbox.get(email) ?? null;
  }
}
